File: rust/day21_part1_2023.rs

use std::fs;

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
struct Coord {
    x: i32,
    y: i32,
}

impl Coord {
    fn add(&self, other: Coord) -> Coord {
        Coord {
            x: self.x + other.x,
            y: self.y + other.y,
        }
    }
}

struct Grid {
    width: i32,
    height: i32,
    data: std::collections::HashMap<Coord, char>,
}

const NORTH: Coord = Coord { x: 0, y: -1 };
const WEST: Coord = Coord { x: -1, y: 0 };
const SOUTH: Coord = Coord { x: 0, y: 1 };
const EAST: Coord = Coord { x: 1, y: 0 };

const EMPTY: char = '.';
const ROCK: char = '#';
const START: char = 'S';

fn is_in_bounds(grid: &Grid, coord: Coord) -> bool {
    0 <= coord.x && coord.x < grid.width && 0 <= coord.y && coord.y < grid.height
}

fn parse_input(input: Vec<String>) -> Grid {
    let mut data = std::collections::HashMap::new();

    for (y, line) in input.iter().enumerate() {
        for (x, char) in line.chars().enumerate() {
            if char != EMPTY {
                data.insert(Coord { x: x as i32, y: y as i32 }, char);
            }
        }
    }

    Grid {
        width: input[0].len() as i32,
        height: input.len() as i32,
        data,
    }
}

fn find_start(grid: &Grid) -> Coord {
    for (coord, &char) in &grid.data {
        if char == START {
            return *coord;
        }
    }
    panic!("No start found.");
}

fn neighbors4(grid: &Grid, coord: Coord) -> Vec<Coord> {
    let neighbors = vec![
        coord.add(NORTH),
        coord.add(SOUTH),
        coord.add(EAST),
        coord.add(WEST),
    ];

    let mut valid_neighbors = vec![];

    for neighbor in neighbors {
        if is_in_bounds(grid, neighbor) && grid.data.get(&neighbor) != Some(&ROCK) {
            valid_neighbors.push(neighbor);
        }
    }

    valid_neighbors
}

fn breadth_first_search(grid: &Grid, start: Coord, neighbor_func: fn(&Grid, Coord) -> Vec<Coord>) -> std::collections::HashMap<Coord, i32> {
    let mut frontier = vec![start];
    let mut reached = std::collections::HashMap::new();
    reached.insert(start, ());
    let mut came_from = std::collections::HashMap::new();
    came_from.insert(start, start);
    let mut distances = std::collections::HashMap::new();
    distances.insert(start, 0);

    while !frontier.is_empty() {
        let current = frontier.remove(0);

        for next in neighbor_func(grid, current) {
            if !reached.contains_key(&next) {
                frontier.push(next);
                reached.insert(next, ());
                came_from.insert(next, current);
                distances.insert(next, distances[&current] + 1);
            }
        }
    }

    distances
}

fn distances_from_extremities(grid: &Grid) -> std::collections::HashMap<Coord, std::collections::HashMap<Coord, i32>> {
    let mut distances = std::collections::HashMap::new();

    let extremities = vec![
        Coord { x: 0, y: 0 },
        Coord { x: grid.width / 2, y: 0 },
        Coord { x: grid.width, y: 0 },
        Coord { x: grid.width, y: grid.height / 2 },
        Coord { x: grid.width, y: grid.height },
        Coord { x: grid.width / 2, y: grid.height },
        Coord { x: 0, y: grid.height },
        Coord { x: 0, y: grid.height / 2 },
    ];

    for &start in &extremities {
        distances.insert(start, breadth_first_search(grid, start, neighbors4));
    }

    distances
}

fn neighbors8(grid: &Grid, coord: Coord) -> Vec<Coord> {
    let neighbors = vec![
        coord.add(NORTH),
        coord.add(SOUTH),
        coord.add(EAST),
        coord.add(WEST),
        coord.add(NORTH).add(EAST),
        coord.add(NORTH).add(WEST),
        coord.add(SOUTH).add(EAST),
        coord.add(SOUTH).add(WEST),
    ];

    neighbors
}

fn solve(input: Vec<String>, num_steps: i32) -> i32 {
    let grid = parse_input(input);

    let start = find_start(&grid);
    let distances = breadth_first_search(&grid, start, neighbors4);

    let mut cnt = 0;
    for &dist in distances.values() {
        if dist <= num_steps && dist % 2 == 0 {
            cnt += 1;
        }
    }
    cnt
}

fn read_file(file_name: &str) -> Vec<String> {
    let content = fs::read_to_string(file_name).expect("Error reading the file");
    content.trim().split('\n').map(|s| s.to_string()).collect()
}

fn main() {
    let input = read_file("input.txt");
    println!("{}", solve(input, 64));
}

File: go/day21_part1_2023.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Coord struct {
	X int
	Y int
}

func (c1 Coord) Add(c2 Coord) Coord {
	return Coord{c1.X + c2.X, c1.Y + c2.Y}
}

func (c Coord) MultiplyByScalar(s int) Coord {
	return Coord{c.X * s, c.Y * s}
}

type Grid struct {
	Width  int
	Height int
	Data   map[Coord]byte
}

var (
	North = Coord{0, -1}
	West  = Coord{-1, 0}
	South = Coord{0, 1}
	East  = Coord{1, 0}
)

const (
	Empty byte = '.'
	Rock  byte = '#'
	Start byte = 'S'
)

func (grid Grid) toString() string {
	res := ""
	for y := 0; y < grid.Height; y++ {
		for x := 0; x < grid.Width; x++ {
			if v, ok := grid.Data[Coord{x, y}]; ok {
				res += string(v)
			} else {
				res += string(Empty)
			}
		}
		res += "\n"
	}
	return res
}

func isInBounds(grid Grid, coord Coord) bool {
	return 0 <= coord.X && coord.X < grid.Width && 0 <= coord.Y && coord.Y < grid.Height
}

func parseInput(input []string) Grid {
	grid := Grid{
		Width:  len(input[0]),
		Height: len(input),
		Data:   make(map[Coord]byte),
	}

	for y, line := range input {
		for x, char := range line {
			if byte(char) != Empty {
				grid.Data[Coord{x, y}] = byte(char)
			}
		}
	}

	return grid
}

func findStart(grid Grid) Coord {
	for coord, char := range grid.Data {
		if char == Start {
			return coord
		}
	}
	panic("No start found.")
}

func neighbors4(grid Grid, coord Coord) []Coord {
	neighbors := []Coord{
		coord.Add(North),
		coord.Add(South),
		coord.Add(East),
		coord.Add(West),
	}

	validNeighbors := []Coord{}

	for _, neighbor := range neighbors {
		if isInBounds(grid, neighbor) && grid.Data[neighbor] != Rock {
			validNeighbors = append(validNeighbors, neighbor)
		}
	}

	return validNeighbors
}

func breadthFirstSearch(grid Grid, start Coord, neighborFunc func(Grid, Coord) []Coord) map[Coord]int {
	frontier := []Coord{start}
	reached := map[Coord]struct{}{start: {}}
	cameFrom := map[Coord]Coord{start: start}
	distances := map[Coord]int{start: 0}

	for len(frontier) > 0 {
		current := frontier[0]
		frontier = frontier[1:]

		for _, next := range neighborFunc(grid, current) {
			if _, ok := reached[next]; !ok {
				frontier = append(frontier, next)
				reached[next] = struct{}{}
				cameFrom[next] = current
				distances[next] = distances[current] + 1
			}
		}
	}

	return distances
}

func distancesFromExtremities(grid Grid) map[Coord]map[Coord]int {
	distances := make(map[Coord]map[Coord]int, 8)

	extremities := []Coord{
		{0, 0},
		{grid.Width / 2, 0},
		{grid.Width, 0},
		{grid.Width, grid.Height / 2},
		{grid.Width, grid.Height},
		{grid.Width / 2, grid.Height},
		{0, grid.Height},
		{0, grid.Height / 2},
	}

	for _, start := range extremities {
		distances[start] = breadthFirstSearch(grid, start, neighbors4)
	}

	return distances
}

func neighbors8(grid Grid, coord Coord) []Coord {
	neighbors := []Coord{
		coord.Add(North),
		coord.Add(South),
		coord.Add(East),
		coord.Add(West),
		coord.Add(North).Add(East),
		coord.Add(North).Add(West),
		coord.Add(South).Add(East),
		coord.Add(South).Add(West),
	}

	return neighbors
}

func solve(input []string, numSteps int) int {
	grid := parseInput(input)

	start := findStart(grid)
	distances := breadthFirstSearch(grid, start, neighbors4)

	cnt := 0
	for _, dist := range distances {
		if dist <= numSteps && dist%2 == 0 {
			cnt++
		}
	}
	return cnt
}

func readFile(fileName string) []string {
	file, err := os.ReadFile(fileName)
	if err != nil {
		panic(err)
	}

	return strings.Split(strings.TrimSpace(string(file)), "\n")
}

func main() {
	input := readFile("input.txt")
	fmt.Println(solve(input, 64))
}